doc: Add pelion user guide (#4748)
nrfconnect/sdk-nrf: doc/nrf/ug_pelion.rst

Pelion Device Management library configuration
==============================================

The Pelion Device Management library features can be controled using Kconfig options.
These options are defined within the Pelion Device Management library repository.

Below is the list of some of the important ones:
* :option:`CONFIG_PELION_UPDATE` - enables device firmware update (DFU) feature.
* :option:`CONFIG_PELION_TRANSPORT_MODE_TCP`, :option:`CONFIG_PELION_TRANSPORT_MODE_UDP`, :option:`CONFIG_PELION_TRANSPORT_MODE_UDP_QUEUE`- choice selecting the transport protocol used by the library.

For more information and a full list of the library configuration options, read the `Zephyr integration tutorial`_ in the Pelion documentation.

Mbed TLS
========

The Pelion Device Management library is depending on a properly configured Mbed TLS library.
To simplify the development within the |NCS|, a predefined set of Mbed TLS configuration options was prepared.
Set compatible with :ref:`nrfxlib:nrf_security` can be enabled using :option:`CONFIG_PELION_NRF_SECURITY` configuration option.

Make sure to properly configure memory region used by Mbed TLS library for a dynamic allocations.

> Memory for this library is allocated from either the libc heap or a dedicated buffer, depending on a configuration.

Reviewer: As discussed f2f: documentation page refs for these options are missing.
Can we mention the Kconfigs for the configuration options?

Author: expanded somehow

   For details, refer to the `Pelion Device Management documentation`_ and `Pelion Device Management release notes`_.

Flash memory layout
===================

The Pelion Device Management library requires that a partition named ``pelion_storage`` is defined on flash.
The partition is used to store the non-volatile data like credentials and identifiers.
Since data kept are fragile it is recommended to keep the partition on the SoC internal flash or otherwise secure access to the memory.

Dynamic allocation memory
=========================

The Pelion Device Management library depends on the new libc.
In case of targets without the memory management unit (MMU), given that the application is not running in the Userspace, the portion of RAM that remains unallocated for any other purpose is used by the libc heap.
Make sure to leave sufficient space for heap so the Pelion Device Management library may work correctly.

Network backend
===============

The Pelion Device Management library uses standard POSIX sockets for the network communication.
It was tested with the following network backends:

> * LTE using the :ref:`lte_lc_readme`

Reviewer: Discussed f2f: The link provided here is not optimal. We either need to find a better one or expand the docs.

Author: expanded somehow

The |NCS| provides Pelion's ``mbed-cloud-client`` library (the Pelion Device Management library) as one of the external submodule repositories managed with :ref:`zephyr:west`.
This library is fetched into the :file:`ncs/modules/lib/pelion-dm` directory of the |NCS|.
Enabling the library allows you to connect to the Pelion cloud platform and exchange data with it using either LTE or OpenThread.

To read more about Pelion, see the `Pelion website`_ and the `Pelion Device Management documentation`_.

An example of the application that is using the Pelion Device Management library in the |NCS| is the :ref:`pelion_client`.

Configuring Pelion Device Management library
********************************************

To use the Pelion Device Management library, you need a developer account on the `Pelion Device Management Portal`_.

You can enable the Pelion Device Management library in the |NCS| with the :option:`CONFIG_PELION_CLIENT` Kconfig option.

Additionally, for the library to work with your device, you must complete the following configuration steps:

1. `Enable and configure Mbed TLS`_.

> #. `Provision the device with credentials`_, required for the connection to the Pelion Device Management Portal.

Reviewer 2: It's without the definite article in the Pelion docs.

#. `Provision the device with update resources`_ for the firmware update procedure to work.
#. `Create a dedicated patition on flash`_ for the Pelion Device Management library
#. `Configure dynamic allocation memory`_ with sufficient space for heap.
#. `Enable and configure the network backend`_.

See the following sections for details.

Enable and configure Mbed TLS
=============================

The Pelion Device Management library depends on a properly configured Mbed TLS library.
To simplify the development within the |NCS|, you can use a predefined set of Mbed TLS configuration options.

You can enable the set that is compatible with :ref:`nrfxlib:nrf_security` using the :option:`CONFIG_PELION_NRF_SECURITY` Kconfig option.

> Make sure to properly configure memory region used by Mbed TLS library for dynamic allocations.

Reviewer 2: ```suggestion
Make sure to properly configure the memory region used by Mbed TLS library for dynamic allocations.
```

> Memory for this library is allocated from either the libc heap or a dedicated buffer, depending on a configuration.

Reviewer 2: ```suggestion
Memory for this library is allocated from either the libc heap or a dedicated buffer, depending on the configuration.
```

> For more information about using the dedicated bugger for Mbed TLS heap please refer to :option:`CONFIG_MBEDTLS_ENABLE_HEAP` and :option:`CONFIG_MBEDTLS_HEAP_SIZE` help.

Reviewer 2: ```suggestion
For more information about using the dedicated buffer for Mbed TLS heap, see :option:`CONFIG_MBEDTLS_ENABLE_HEAP` and :option:`CONFIG_MBEDTLS_HEAP_SIZE` help.
```

> If :option:`CONFIG_MBEDTLS_ENABLE_HEAP` is disabled, if Mbed TLS configuration files do not add any overrides, the libc heap is used.

Reviewer 2: ```suggestion
If :option:`CONFIG_MBEDTLS_ENABLE_HEAP` is disabled and Mbed TLS configuration files do not add any overrides, the libc heap is used.
```

Provision the device with credentials
=====================================

> You must provision the device before it can connect with the Pelion Device Management Portal.

Reviewer 2: ```suggestion
You must provision the device before it can connect to Pelion Device Management Portal.
```

When using Pelion, you can complete the provisioning process using either a production tool (Factory Configurator Utility for factory provisioning) or the developer flow process.
For more information about the provisioning process, see `Provisioning devices for Pelion Device Management`_ in the Pelion documentation.

Provision the device with update resources
==========================================

For details about provisioning the device with update resources, see the `Device Management Update`_ guide in the Pelion documentation.

.. note::
   Zephyr's port of Pelion might not support the most recent update client features.
   The Device Management Update link points to the documentation of the version compatible with the |NCS|.
   For details, refer to the `Pelion Device Management documentation`_ and `Pelion Device Management release notes`_.

> Create a dedicated patition on flash

Reviewer 2: ```suggestion
Create a dedicated partition on flash
```
Also extend the underlining under the heading.

====================================

The Pelion Device Management library requires a partition named ``pelion_storage`` to be defined on flash.
The partition is used to store the non-volatile data, such as credentials and identifiers.
Because this kind of data is fragile, keep the partition on the SoC internal flash or make sure to secure the access to the memory.

Configure dynamic allocation memory
===================================

The Pelion Device Management library depends on the new libc.
In case of targets without the memory management unit (MMU), given that the application is not running in the Userspace, the portion of RAM that remains unallocated for any other purpose is used by the libc heap.

> Make sure to leave a sufficient space for heap so the Pelion Device Management library may work correctly.

Reviewer 2: ```suggestion
Make sure to leave enough space for the heap so that the Pelion Device Management library works correctly.
```

> For more information about C standard library refer to :ref:`libc_api`.

Reviewer 2: ```suggestion
For more information about C standard library, see the :ref:`libc_api`.
```

Enable and configure the network backend
========================================

The Pelion Device Management library uses standard POSIX sockets for the network communication.
The library was tested with the following network backends:

* Cellular (LTE, NB-IoT)
* OpenThread

> After initializing and setting up the Pelion object instance, it continuously retries to connect to the Pelion's Device Management server.

Reviewer 2: ```suggestion
After initializing and setting up the Pelion object instance, it continuously retries to connect to Pelion's Device Management server.
```

Cellular backend
----------------

> Cellular backend makes use of the on board modem being part of the nNR9160 SiP.

Reviewer 2: ```suggestion
The cellular backend uses the on-board modem that is part of the nRF9160 SiP.
```

> For more information about working with the modem please refer to :ref:`ug_nrf9160`.

Reviewer 2: ```suggestion
For more information about working with the modem, see the :ref:`ug_nrf9160` documentation.
```

> To simplify the connection setup the LTE link controller can be used (see :ref:`lte_lc_readme`).

Reviewer 2: ```suggestion
To simplify the connection setup, you can use the LTE link controller (see :ref:`lte_lc_readme`).
```

OpenThread backend
------------------

Thread is a low-power mesh networking technology.

> I allows a device to access the Internet if one of the mesh elements (boarder router) share the Internet connection.

Reviewer 2: ```suggestion
It allows a device to access the Internet if one of the mesh elements (boarder router) share the Internet connection.
```

> For more information please refer to :ref:`zephyr:thread_protocol_interface` and :ref:`ug_thread`.

Reviewer 2: ```suggestion
For more information, see :ref:`zephyr:thread_protocol_interface` and :ref:`ug_thread`.
```

Additional configuration
************************

You can control the Pelion Device Management library features using Kconfig options that are defined within the Pelion Device Management library repository.

The following options are among the most important ones:

* :option:`CONFIG_PELION_UPDATE` - This option enables the device firmware update (DFU) feature.
* :option:`CONFIG_PELION_TRANSPORT_MODE_TCP`, :option:`CONFIG_PELION_TRANSPORT_MODE_UDP`, :option:`CONFIG_PELION_TRANSPORT_MODE_UDP_QUEUE` - These options select the transport protocol used by the library.

> For more information and the full list of the library configuration options, read the `Zephyr integration tutorial`_ in the Pelion documentation.

Reviewer 2: Looks like the link is not valid (https://developer.pelion.com/docs/device-management/current/connecting/zephyr.html). It gives me 404. Also searching for Zephyr in the Pelion docs site did not give any results.

Reviewer: This is the link provided by @marcuschangarm and it will be populated by the release time.
